rule.go

```
@@ -31,6 +31,34 @@ type Rule struct {
 	Type              uint8
 }
 
+func (r Rule) Equal(x Rule) bool {
+	return r.Table == x.Table &&
+		((r.Src == nil && x.Src == nil) ||
+			(r.Src != nil && x.Src != nil && r.Src.String() == x.Src.String())) &&
+		((r.Dst == nil && x.Dst == nil) ||
+			(r.Dst != nil && x.Dst != nil && r.Dst.String() == x.Dst.String())) &&
+		r.OifName == x.OifName &&
+		r.Priority == x.Priority &&
+		r.Family == x.Family &&
+		r.IifName == x.IifName &&
+		r.Invert == x.Invert &&
+		r.Tos == x.Tos &&
+		r.Type == x.Type &&
+		r.IPProto == x.IPProto &&
+		r.Protocol == x.Protocol &&
+		r.Mark == x.Mark &&
+		(ptrEqual(r.Mask, x.Mask) || (r.Mark != 0 &&
+			(r.Mask == nil && *x.Mask == 0xFFFFFFFF || x.Mask == nil && *r.Mask == 0xFFFFFFFF))) &&
```

up till here extracted from rule_test.go

---

I don't understand the semantics of comparing the masks with `0xFFFFFFFF`. Shouldn't the masks be identical for the rules to match?

---

Let me try to walk you through line 50. Not my code. I was also puzzled.
1. We already know that the Marks are equal.
2. ptrEqual compares the masks. If they are not equal (both nil, both the same value), madness beginns
 - we treat a nil and an all-bits-set mask as equal
 - and if the Mark is non-zero, we consider the two rules equal with regard to their mark/mask.
No conclusion yet, what this does. In particular, the non-zero check for Mark throws me off.

---

I think we need to know this before we can introduce a generic `Equal()` method. Otherwise, people might relying on the semantics being different from what it actually is.

My guess would be that the purpose is to mask an issue where the value written to the kernel differs from the value read back. That happens sometimes with `netlink`.

---

I found this: https://unix.stackexchange.com/questions/436799/how-the-fwmark-works-together-with-mask-in-ip-rule-command

If mark is non-zero, then the default mask is all-bits-set.
So rule fragments `fwmark 1` and `fwmark 1/0xffffffff` need to be considered equal.

---

https://github.com/torvalds/linux/blob/master/net/core/fib_rules.c#L621

---

I can see where you're coming from with this just being moved code, but I would like a comment with a description + link.
In the best of all worlds, I would also have liked a unit test, but I think its too much to ask, given the current state of the code base.

This is just the kind of code, which might be slightly broken because we only see normalized input data in practice, so it would be nice to have it unit tested.

---

Added comment and improved unit test.